blockchain/blockdao/blob_store.go

```diff
@@ -22,4 +35,173 @@ type (
+	//    bucket as the storage index. It converts the requested blob hash to the
+	//    bucket and block number, and retrieves the blob object from storage
+	blobStore struct {
+		kvStore        db.KVStore
+		indexer        *blobIndexer
+		totalBucket    uint32
+		blockPerBucket uint32
+	}
 )
+
+func NewBlobStore(kv db.KVStore, totalBucket, blockPerBucket uint32) *blobStore {
+	return &blobStore{
+		kvStore:        kv,
+		totalBucket:    totalBucket,
+		blockPerBucket: blockPerBucket,
+	}
+}
+
+func (bs *blobStore) Start(ctx context.Context) error {
+	if err := bs.kvStore.Start(ctx); err != nil {
+		return err
+	}
+	start, err := bs.checkBuckets()
+	if err != nil {
+		return err
+	}
+	bs.indexer = newBlobIndexer(bs.totalBucket, bs.blockPerBucket, start)
+	return nil
+}
+
+func (bs *blobStore) Stop(ctx context.Context) error {
+	if err := bs.kvStore.Stop(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (bs *blobStore) GetBlob(h hash.Hash256) (*types.BlobTxSidecar, string, error) {
+	block, err := bs.indexer.indexByActionHash(h[:])
+	if err != nil {
+		return nil, "", err
+	}
+	blobs, hashes, err := bs.getBlobs(block)
+	if err != nil {
+		return nil, "", err
+	}
+	for i, v := range hashes {
+		if common.BytesToHash(h[:]).Hex() == v {
+			return blobs[i], v, nil
+		}
+	}
+	return nil, "", errors.Wrapf(db.ErrNotExist, "blob hash = %x", h)
+}
+
+func (bs *blobStore) GetBlobsByHeight(height uint64) ([]*types.BlobTxSidecar, []string, error) {
+	return bs.getBlobs(height)
+}
+
+func (bs *blobStore) PutBlock(blk *block.Block) error {
+	if blk.Height() <= bs.indexer.tipHeight() {
+		return errors.Errorf("block height %d is less than current tip height", blk.Height())
+	}
+	pb := iotextypes.BlobTxSidecars{
+		TxHash:   make([][]byte, 0),
+		Sidecars: make([]*iotextypes.BlobTxSidecar, 0),
+	}
+	for _, act := range blk.Actions {
+		if b := act.BlobTxSidecar(); b != nil {
+			h, err := act.Hash()
+			if err != nil {
+				return err
+			}
+			pb.TxHash = append(pb.TxHash, h[:])
+			pb.Sidecars = append(pb.Sidecars, action.ToProtoSideCar(b))
+		}
+	}
+	raw, err := proto.Marshal(&pb)
+	if err != nil {
+		return errors.Wrapf(err, "failed to put block = %d", blk.Height())
+	}
+	return bs.putBlob(raw, blk.Height())
+}
+
+func (bs *blobStore) putBlob(blob []byte, height uint64) error {
+	var (
+		bucket, start, expired = bs.indexer.bucketByHeight(height)
+		ns                     = nsForBucket(bucket)
+		heightKey              = keyForBlock(height)
+		err                    error
+	)
+	if expired {
+		// clear expired bucket
+		if err := bs.kvStore.Delete(ns, nil); err != nil {
+			return errors.Wrapf(err, "failed to delete expired bucket = %s", ns)
+		}
+		b := batch.NewBatch()
+		b.Put(ns, heightKey, blob, fmt.Sprintf("failed to write blob at height %d", height))
+		b.Put(ns, _startKey, keyForBlock(start), fmt.Sprintf("failed to update start height %d", height))
+		err = bs.kvStore.WriteBatch(b)
```

Comment:
`Delete(ns, nil)` deletes the entire bucket. will have another PR to accommodate this into `WriteBatch` so it will a single DB call